### PR TITLE
Add FortCL initialization options (issue #1019)

### LINKED ISSUE #1019: Add FortCL initialization options as options in the OCLTrans transformation
Recently we added a few parameters in FortCL for initializing the OpenCL device. It will be good to expose these as transformation options in PSyclone (the infrastructure already exists for the 'end_barrier' option). 

```
cltrans.apply(schedule, options={"end_barrier": True,
                                 "enable_profiling": False,
                                 "out_of_order": False,
                                 "devices_per_node": 1})
```

### PR DESCRIPTION
**Pull request by sergisiso**
Closes #1019

Adds OpenCL initialization options and updates the OpenCL documentation.

Unfortunately I couldn't see how to do the environment var solution described and preferred in #1019 because in this case the environment var string has to be converted into an integer, probably with the read statement which is not supported in f2pygen, so I went for the configuration file parameter.

It can be tested with the new `compile-mpi-ocl` in examples/gocean/example3 or in PSycloneBench

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter Ready for first review

**Review by arporter — CHANGES_REQUESTED** [review]
Good job Sergi. Just minor comments.
All tests pass, including compilation.
Ditto the examples.
Updated documentation builds fine.
I'll check for pylint/pycodestyle next time.

Please could you update the copyright date in any files you've touched.

File: config/psyclone.cfg [hunk 1]
```diff
@@ -48,6 +48,8 @@ REPRODUCIBLE_REDUCTIONS = false
 REPROD_PAD_SIZE = 8
 PSYIR_ROOT_NAME = psyir_tmp
 VALID_PSY_DATA_PREFIXES = profile extract read_only_verify nan_test
+# Specify number of OpenCL devices per node
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply config/psyclone.cfg:51]: Done

File: doc/user_guide/transformations.rst [hunk 2]
```diff
@@ -708,21 +712,39 @@ multiple InvokeSchedule and kernel-specific optimization options.
...
+OpenCL delays the decision of which and where kernels will execute until
+run-time, therefore it is important to use the environment variables provided
+by FortCL and DL_ESM_INF to inform how things should execute. Specifically:
+
+- ``FORTCL_KERNELS_FILE``: Point to the file containing the kernels to execute,
+  they can be compiled ahead-of-time or providing the source for JIT
+  compilation. To link more than a single kernel, one must merge all the
+  kernels generated by PSyclone in a single source file.
+- ``FORTCL_PLATFORM``: If the system has more than 1 OpenCL platform. This
+  environment variables allows to select which platform to execute the kernels.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/transformations.rst:724]: Done

File: doc/user_guide/transformations.rst [hunk 3]
```diff
@@ -708,21 +712,39 @@ multiple InvokeSchedule and kernel-specific optimization options.
...
+- ``DL_ESM_ALIGNMENT``: When using OpenCL <= 1.2 the local_size should be
+  exactly divisible by the total size. If this is not the case some
+  implementations fail silently. A way to solve this issue is to set the
+  `DL_ESM_ALIGNMENT` variable to be equal to the local size.
+
 
+.. note:: The OpenCL generation can be combined with distributed memory
+    generation. In the case that more than 1 accelerator per node are
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/transformations.rst:732]: Done

File: doc/user_guide/transformations.rst [hunk 4]
```diff
@@ -708,21 +712,39 @@ multiple InvokeSchedule and kernel-specific optimization options.
...
+    available, the PSyclone configuration file variable
+    ``OCL_DEVICES_PER_NODE`` has to be set to the appropriate value and
+    the `mpirun` command have to match this value accordingly.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/transformations.rst:735]: Done

File: doc/user_guide/transformations.rst [hunk 5]
```diff
@@ -708,21 +712,39 @@ multiple InvokeSchedule and kernel-specific optimization options.
...
 
-Because OpenCL kernels are linked at run-time, it will be up to the run-time
-environment to specify which of the kernels to use. For instance, one could
-merge multiple kernels together in a single binary file and
-use the `PSYCLONE_KERNELS_FILE` provided by the FortCL library.
+    For instance if there are 2 accelerator per nodes, `psyclone.cfg` should
+    have `OCL_DEVICES_PER_NODE=2` and the program must be executed with
+    ``mpirun -n <total_ranks> -ppn 2 ./application``
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/transformations.rst:739]: Done

File: doc/user_guide/transformations.rst [hunk 6]
```diff
@@ -708,21 +712,39 @@ multiple InvokeSchedule and kernel-specific optimization options.
...
+
+
+For example, an execution of a PSyclone generated OpenCL code using all the mentioned
+run-time configuration could look something like:
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/user_guide/transformations.rst:743]: Done

File: examples/gocean/eg3/README.md [hunk 7]
```diff
@@ -56,6 +56,24 @@ kernels. To run the application with this file use the following command:
...
 ```
 
+This Makefile also provides a target to combine OpenCL and distributed
+memory:
+```sh
+make compile-mpi-ocl
+```
+
+In addition to the details above, the distributed memory target needs an mpi
+compiler in the F90 environment variable and that the libraries have been
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/gocean/eg3/README.md:66]: Done

File: examples/gocean/eg3/Makefile [hunk 8]
```diff
@@ -67,19 +70,37 @@ ${FORTCL_LIB}:
...
+	@echo "OpenCL compilation complete!"
+	@echo "Use 'FORTCL_KERNELS_FILE=allkernels.cl ./alg_opencl.exe' to run."
+
+compile-mpi-ocl: dm_fd_lib ${FORTCL_LIB} ${KERNELS} psy_gen_dm.o alg_gen_dm.f90
+	${F90} ${LDFLAGS} -I${INF_INC} alg_gen_dm.f90 psy_gen_dm.o ${INF_LIB} \
+		${FORTCL_LIB} -lOpenCL -o alg_dm_opencl.exe
+	@echo "//OpenCL Kernels file" > allkernels.cl
+	cat compute_*_0.cl >> allkernels.cl
+	@echo "OpenCL compilation complete!"
+	@echo "Use 'FORTCL_KERNELS_FILE=allkernels.cl mpirun -n 2" \
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/gocean/eg3/Makefile:94]: > Worth changing "2" to be "<NUM DEVICES>"

Good idea

> saying that it must match the value in the config file?

This is just the simple case where there is a device per node and I think is Ok to just print this. I will add an example of multiple devices per node in the README instead.

File: examples/gocean/eg3/README.md [hunk 9]
```diff
@@ -56,6 +56,24 @@ kernels. To run the application with this file use the following command:
...
+compiled with such compiler. The example can be compiled and executed with
+the following commands:
+
+```sh
+make allclean
+export F90=mpifc
+make compile-mpi-ocl
+FORTCL_KERNELS_FILE=allkernels.cl mpirun -n 2 ./alg_dm_opencl.exe
+```
+
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/gocean/eg3/README.md:76]: Done

File: examples/gocean/eg3/README.md [hunk 10]
```diff
@@ -56,6 +56,24 @@ kernels. To run the application with this file use the following command:
...
 
+This Makefile also provides a target to combine OpenCL and distributed
+memory:
+```sh
+make compile-mpi-ocl
+```
+
+In addition to the details above, the distributed memory target needs an mpi
+compiler in the F90 environment variable and that the libraries have been
+compiled with such compiler. The example can be compiled and executed with
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/gocean/eg3/README.md:67]: Not exactly, the example just uses 2 nodes. I made this clearer and added a second example with multiple devices per node.

File: src/psyclone/configuration.py [hunk 11]
```diff
@@ -329,6 +332,15 @@ def load(self, config_file=None):
         except KeyError:
             self._valid_psy_data_prefixes = []
 
+        try:
+            self._ocl_devices_per_node = self._config['DEFAULT'].getint(
+                'OCL_DEVICES_PER_NODE')
+        except ValueError as err:
+            raise ConfigurationError(
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/configuration.py:339]: Done

File: src/psyclone/gocean1p0.py [hunk 12]
```diff
@@ -176,6 +176,18 @@ def __init__(self, alg_calls, psy):
                     # those seen so far
                     index_offsets.append(kern_call.index_offset)
 
+    def gen_rank_expression(self, scope):
+        ''' Generate the expression to retrieve the process rank.
+
+        :param scope: where the expression is going to be located.
+        :type scope: :py:class:`psyclone.f2pygen.BaseGen`
+        :return: generate the expression to retrieve the process rank.
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/gocean1p0.py:184]: Done

File: src/psyclone/psyGen.py [hunk 13]
```diff
@@ -338,17 +338,47 @@ def gen_code(self, parent):
...
+        def raise_unmatching_options(option_name):
+            ''' Create unmatching OpenCL options error message.
+
+            :param str option_name: name of the option that failed.
+            :raises NotImplementedError: with the given option_name.
+            '''
+            raise NotImplementedError(
+                "The current implementation creates a single OpenCL context "
+                "for all the invokes which needs certain OpenCL options to "
+                "match between invokes. Found '{0}' with unmathcing values "
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py:350]: Done

File: src/psyclone/psyGen.py [hunk 14]
```diff
@@ -359,10 +389,11 @@ def gen_code(self, parent):
...
             invoke.gen_code(parent)
         if generate_ocl_init:
-            self.gen_ocl_init(parent, opencl_kernels, opencl_num_queues)
+            self.gen_ocl_init(parent, opencl_kernels, opencl_num_queues,
+                              ocl_enable_profiling, ocl_out_of_order)
 
-    @staticmethod
-    def gen_ocl_init(parent, kernels, num_queues):
+    def gen_ocl_init(self, parent, kernels, num_queues, enable_profiling,
+                     out_of_order):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py:396]: Added docstrings

File: src/psyclone/psyGen.py [hunk 15]
```diff
@@ -414,10 +472,22 @@ def gen_ocl_init(parent, kernels, num_queues):
...
+
+        :param scope: where the expression is going to be located.
+        :type scope: :py:class:`psyclone.f2pygen.BaseGen`
+        :return: generate the expression to retrieve the process rank.
+        :rtype: str
+        :raises NotImplementedError: this is an abstract method.
+
+        '''
+        raise NotImplementedError("Abstract method")
+
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py:490]: Done

File: src/psyclone/tests/config_test.py [hunk 16]
```diff
@@ -63,6 +63,7 @@
 REPRODUCIBLE_REDUCTIONS = false
 REPROD_PAD_SIZE = 8
 VALID_PSY_DATA_PREFIXES = profile extract
+OCL_DEVICES_PER_NODE = 1
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/config_test.py:66]: Done

File: src/psyclone/tests/gocean1p0_opencl_test.py [hunk 17]
```diff
@@ -322,25 +322,69 @@ def test_psy_init(kernel_outputdir):
...
+        "        CALL ocl_env_init(5, ocl_device_num, .False., .False.)\n"
+        "        ! The kernels this PSy layer module requires\n"
+        "        kernel_names(1) = \"compute_cu_code\"\n"
+        "        ! Create the OpenCL kernel objects. Expects to find all of "
+        "the compiled\n"
+        "        ! kernels in FORTCL_KERNELS_FILE.\n"
+        "        CALL add_kernels(1, kernel_names)\n"
+        "      END IF\n"
+        "    END SUBROUTINE psy_init\n")
+    print(generated_code)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/gocean1p0_opencl_test.py:368]: Ups.. removed

File: src/psyclone/tests/psyGen_test.py [hunk 18]
```diff
@@ -222,6 +222,13 @@ def test_valid_return_object_from_name():
     transform = trans.get_trans_name("LoopFuse")
     assert isinstance(transform, Transformation)
 
+# Test Invokes abstract methods
+
+def test_invokes_abstract_methods():
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyGen_test.py:227]: Deleted test

File: src/psyclone/tests/gocean1p0_opencl_test.py [hunk 19]
```diff
@@ -381,6 +437,27 @@ def test_opencl_options_validation():
         in str(err.value)
 
 
+@pytest.mark.usefixtures("kernel_outputdir")
+@pytest.mark.parametrize("option_to_check", ['enable_profiling',
+                                             'out_of_order'])
+def test_opencl_multi_invoke_options_validation(option_to_check):
+    ''' Check that the OpenCL options constrains when there are multiple
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/gocean1p0_opencl_test.py:444]: Done

File: src/psyclone/tests/gocean1p0_opencl_test.py [hunk 20]
```diff
@@ -362,6 +406,18 @@ def test_opencl_options_validation():
     assert "InvokeSchedule opencl_option 'end_barrier' should be a boolean." \
         in str(err.value)
 
+    # enable_profiling option must be a boolean
+    with pytest.raises(TransformationError) as err:
+        otrans.apply(sched, options={'enable_profiling': 1})
+    assert ("InvokeSchedule opencl_option 'enable_profiling' should be a "
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/gocean1p0_opencl_test.py:412]: Changed this and a few similar error messages to "OpenCL option"

File: src/psyclone/tests/gocean1p0_opencl_test.py [hunk 21]
```diff
@@ -322,25 +322,69 @@ def test_psy_init(kernel_outputdir):
...
+    OpenCL environment with the provided non-default options. '''
+    psy, _ = get_invoke("single_invoke.f90", API, idx=0)
+    sched = psy.invokes.invoke_list[0].schedule
+    otrans = OCLTrans()
+    otrans.apply(sched, options={"end_barrier": True,
+                                 "enable_profiling": True,
+                                 "out_of_order": True})
+    generated_code = str(psy.gen)
+    assert "CALL ocl_env_init(1, ocl_device_num, .True., .True.)\n" \
+        in generated_code
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/gocean1p0_opencl_test.py:385]: Sure

File: src/psyclone/psyGen.py [hunk 22]
```diff
@@ -414,10 +472,22 @@ def gen_ocl_init(parent, kernels, num_queues):
         ifthen.add(CommentGen(ifthen,
                               " Create the OpenCL kernel objects. Expects "
                               "to find all of the compiled"))
-        ifthen.add(CommentGen(ifthen, " kernels in PSYCLONE_KERNELS_FILE."))
+        ifthen.add(CommentGen(ifthen, " kernels in FORTCL_KERNELS_FILE."))
         ifthen.add(CallGen(ifthen, "add_kernels", [nkernstr, "kernel_names"]))
 
 
+    def gen_rank_expression(self, scope):
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/psyGen.py:479]: Ah yes, I always forget the abc library

File: examples/gocean/eg3/alg.f90 [hunk 23]
```diff
@@ -65,8 +65,8 @@ program simple
 
   ! Create our grid
   model_grid = grid_type(GO_ARAKAWA_C,                                 &
-                         (/GO_BC_PERIODIC,GO_BC_PERIODIC,GO_BC_NONE/), &
-                         GO_OFFSET_SW)
+                         (/GO_BC_EXTERNAL,GO_BC_EXTERNAL,GO_BC_NONE/), &
+                         GO_OFFSET_NE)
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/gocean/eg3/alg.f90:69]: I changed it because with "GO_OFFSET_SW" it fails with:  `cv_sw_init: IMPLEMENT non-periodic BCs!`
I guess this is the limitation of dl_esm_inf described here: https://github.com/stfc/dl_esm_inf/issues/46

I now updated the kernels 'index_offset' also to GO_OFFSET_NE but I don't know if this is good enough or it is not as simple as just changing the kernel label.

**Comment by sergisiso** [comment]
@arporter PR ready for next review

File: doc/user_guide/transformations.rst [hunk 24]
```diff
@@ -721,30 +721,33 @@ by FortCL and DL_ESM_INF to inform how things should execute. Specifically:
...
-    generation. In the case that more than 1 accelerator per node are
-    available, the PSyclone configuration file variable
+    generation. In the case where there is more than one accelerator available
+    on each node, the PSyclone configuration file parameter
     ``OCL_DEVICES_PER_NODE`` has to be set to the appropriate value and
-    the `mpirun` command have to match this value accordingly.
+    the number of MPI-ranks-per-node set by the `mpirun` command has to match
+    this value accordingly.
 
     For instance if there are 2 accelerator per nodes, `psyclone.cfg` should
```
Comment (arporter) [inline doc/user_guide/transformations.rst:739]: "accelerators per node"

File: doc/user_guide/transformations.rst [hunk 25]
```diff
@@ -721,30 +721,33 @@ by FortCL and DL_ESM_INF to inform how things should execute. Specifically:
...
-    have `OCL_DEVICES_PER_NODE=2` and the program must be executed with
-    ``mpirun -n <total_ranks> -ppn 2 ./application``
+    have ``OCL_DEVICES_PER_NODE=2`` and the program must be executed with
+    ``mpirun -n <total_ranks> -ppn 2 ./application`` (Note: `-ppn` is an
+    Intel MPI specific parameter, use equivalet configuration parameters for
```
Comment (arporter) [inline doc/user_guide/transformations.rst:742]: "equivalent"

File: examples/gocean/eg3/Makefile [hunk 26]
```diff
@@ -1,7 +1,7 @@
 # -----------------------------------------------------------------------------
 # BSD 3-Clause License
 #
-# Copyright (c) 2020, Science and Technology Facilities Council.
+# Copyright (c) 2021, Science and Technology Facilities Council.
```
Comment (arporter) [inline examples/gocean/eg3/Makefile:4]: Strictly 2020-2021 (because this file was already present I think?)

File: examples/gocean/eg3/Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (arporter) [reply examples/gocean/eg3/Makefile]: Thanks.

File: examples/gocean/eg3/README.md [hunk 27]
```diff
@@ -74,13 +74,29 @@ make compile-mpi-ocl
 FORTCL_KERNELS_FILE=allkernels.cl mpirun -n 2 ./alg_dm_opencl.exe
 ```
 
+The previous example uses a single accelerator device in each node. If more
+than one accelerator is available in each node, these can be used by setting
+the `OCL_DEVICES_PER_NODE` parameter in the PSyclone configuration file and
+the providing the mpirun command with the appropriate parameter to place as
+many MPI ranks per node as accelerator devices the node has. For example,
```
Comment (arporter) [inline examples/gocean/eg3/README.md:81]: "...per node as there are accelerator devices."

File: examples/gocean/eg3/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (arporter) [reply examples/gocean/eg3/README.md]: Thanks for clarifying.

File: examples/gocean/eg3/alg.f90 (hunk 23 above)
[the remark this answers is not in the document]
Comment (arporter) [reply examples/gocean/eg3/alg.f90:69]: Unfortunately it's not quite as simple as that. I can't remember the details but it will require a shift in the array indices, maybe something like `i - 1` -> `i` for accesses to fields on U pts, `j - 1` -> `j` on V pts and both on F pts. See the picture in the https://psyclone.readthedocs.io/en/stable/gocean1p0.html#grid section. Since this is just an example the simplest solution may be to simply replace this kernel with one from `shallow` in PSycloneBench since that uses SW offset.

[the remark this answers is not in the document]
Comment (arporter) [reply examples/gocean/eg3/alg.f90:69]: Alternatively, you could try keeping GO_BC_PERIODIC although that might cause other problems?

File: examples/gocean/eg3/compute_cu_mod.f90 [hunk 28]
```diff
@@ -63,7 +63,7 @@ module compute_cu_mod
      !! point. This kernel assumes that the U,V and F points that
      !! share the same index as a given T point are those immediately
      !! to the South and West of it.
-     integer :: index_offset = GO_OFFSET_SW
```
Comment (arporter) [inline examples/gocean/eg3/compute_cu_mod.f90:66]: As already said, we can't change this without changing the kernel implementation.

File: src/psyclone/tests/gocean1p0_opencl_test.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/gocean1p0_opencl_test.py]: Thanks, that's nicer.

File: src/psyclone/tests/psyGen_test.py [hunk 29]
```diff
@@ -1,7 +1,7 @@
 # -----------------------------------------------------------------------------
 # BSD 3-Clause License
 #
-# Copyright (c) 2017-2020, Science and Technology Facilities Council.
```
Comment (arporter) [inline src/psyclone/tests/psyGen_test.py:4]: Can you revert all changes to this file?

**Review by arporter — CHANGES_REQUESTED** [review]
Apart from the problem of the offset used in the example kernels, we're down to very minor things now.
All examples are OK, incl. with compilation.
All tests pass, incl. with compilation.
Updated user guide builds fine.
Ref guide builds without new warnings.
I've checked all files for conformance.

re. the kernel-offset issue, I realise I've lost sight of the root cause of this problem. Did you have to change the boundary conditions on the grid in order to solve an execution problem? As this is only an example, I don't think it's worth putting in lots of effort, we just need a simple solution (if there is one).

File: doc/user_guide/transformations.rst (hunk 24 above)
Comment (sergisiso) [reply doc/user_guide/transformations.rst:739]: Fixed

File: doc/user_guide/transformations.rst (hunk 25 above)
Comment (sergisiso) [reply doc/user_guide/transformations.rst:742]: Fixed

File: examples/gocean/eg3/Makefile (hunk 26 above)
Comment (sergisiso) [reply examples/gocean/eg3/Makefile:4]: Ups.. good catch

File: examples/gocean/eg3/README.md (hunk 27 above)
Comment (sergisiso) [reply examples/gocean/eg3/README.md:81]: Done

File: src/psyclone/tests/psyGen_test.py (hunk 29 above)
Comment (sergisiso) [reply src/psyclone/tests/psyGen_test.py:4]: Done

**Comment by sergisiso** [comment]
> re. the kernel-offset issue, I realise I've lost sight of the root cause of this problem. Did you have to change the boundary conditions on the grid in order to solve an execution problem? As this is only an example, I don't think it's worth putting in lots of effort, we just need a simple solution (if there is one).

@arporter The reason for the changes are limitations on dl_esm_inf distributed memory.  Both BC_PERIODIC and the GO_OFFSET_SW stop in the initialization step.

I have now tried to just change to BC_EXTERNAL and comment out the dl_esm_inf lines in the offset_sw initialization:
field_mod.f90:884:
```
! I think these are correct but we stop because I've not properly
! gone through the coding.
call gocean_stop('cf_sw_init: CHECK non-periodic BCs!')
```
This already solves the issue with this example, but I don't quite understand what  'gone through the coding' means here? Does it mean that it need proper tests to guarantee that it works?

Ultimately I would like to solve the BC_PERIODIC as well because it is needed for the Shallow benchmark. Which in turn could be used as a test that everything works as expected.

So the options we have is:
1) Completely change the example to an OFFSET_NE code
2) Block the MPI Makefile with a TODO to the missing dl_esm_inf functionality (I am ok with this if its temporal as I am already testing it with NemoLite2D)
3) Make a quick PR that comments out the sw_init non-periodic BC guard
4) Implement BC_PERIODIC and delete the sw_init guard that then can be tested against the Shallow serial version

I can not assess how safe is 3 or how long would it take to do 4. So I will appreciate your input on this.

**Comment by arporter** [comment]
I think option 2. is the best bet with a new PR in dl_esm_inf to develop/test that the necessary functionality works. "By gone through the code" I think I meant I haven't  self-reviewed it to check it. If I was a betting man I would bet that it is probably not quite right.

File: examples/gocean/eg3/alg.f90 (hunk 23 above)
[the remark this answers is not in the document]
Comment (sergisiso) [reply examples/gocean/eg3/alg.f90]: Reverted the changes and added a TODO to the relevant dl_esm_inf issue.

File: examples/gocean/eg3/compute_cu_mod.f90 (hunk 28 above)
Comment (sergisiso) [reply examples/gocean/eg3/compute_cu_mod.f90]: Reverted the changes and added a TODO to the relevant dl_esm_inf issue.

**Comment by sergisiso** [comment]
@arporter Ready for next review.

**Review by arporter — APPROVED** [review]
